File: HaskellCocoa/Cocoa/AppKit/HNSMenuItem.m
//
//  HNSObject.m
//  
//
//  Created by middleware on 2/13/16.
//
//

#import <AppKit/AppKit.h>

#import "HNSObject.h"


typedef void (*HsAction)(HsPtr menuItem, HsPtr target);


@interface HNSMenuItemTarget : NSObject

@property (assign, nonatomic) HsFunPtr hsAction;
@property (assign, nonatomic) HsPtr hsTarget;
@property (strong, nonatomic) id menuItemTarget;

- (void)actionFunc:(id)sender;

@end



@implementation HNSMenuItemTarget

- (void)actionFunc:(id)sender
{
    HsAction pAction = (HsAction)_hsAction;
    pAction((__bridge HsPtr)sender, _hsTarget);
}

- (void)dealloc
{
    hs_free_fun_ptr(_hsAction);
}

@end




void hns_menuItem_removeAction(HsPtr menuItem);

void hns_menuItem_setAction(HsPtr menuItem, HsPtr hsTarget, HsFunPtr action)
{
    hns_menuItem_removeAction(menuItem);
    
    NSMenuItem* pItem = (__bridge NSMenuItem*)menuItem;
    HNSMenuItemTarget* target = [HNSMenuItemTarget new];
    target.hsAction = action;
    target.hsTarget = hsTarget;
    hns_retain(target);
    
    pItem.target = target;
    pItem.action = @selector(actionFunc:);
}


void hns_menuItem_removeAction(HsPtr menuItem)
{
    NSMenuItem* pItem = (__bridge NSMenuItem*)menuItem;
    id oldTarget = pItem.target;
    
    pItem.target = nil;
    pItem.action = nil;
    
    if (oldTarget)
        hns_release(oldTarget);
}

File: HaskellCocoa/Cocoa/AppKit/HNSMenuItem.hs



{-# LANGUAGE ForeignFunctionInterface #-}



module Cocoa.AppKit.HNSMenuItem
(
    HNSMenuItem
,   HNSMenuItemObj
,   nsMenuItem_setAction
,   nsMenuItem_removeAction
)
where



import Foreign




class (HNSObject a) => HNSMenuItem a where

    nsMenuItem_setAction :: Ptr a ->                            -- sender                               
                            FunPtr (Ptr a -> IO ()) ->          -- action
                            IO ()

    nsMenuItem_setAction = hns_menuItem_setAction


    nsMenuItem_removeAction :: Ptr a -> IO ()

    nsMenuItem_removeAction = hns_menuItem_removeAction






data HNSMenuItemObj;

instance HNSObject HNSMenuItemObj where
instance HNSMenuItem HNSMenuItemObj where



foreign import ccall hns_menuItem_setAction :: Ptr a -> FunPtr (Ptr a -> IO ()) -> IO ()
foreign import ccall hns_menuItem_removeAction :: Ptr a -> IO ()






